File: leetcode/0033-search-in-rotated-sorted-array.cc
// https://leetcode.com/problems/search-in-rotated-sorted-array/
// Medium

#include<vector>

using namespace std;

class Solution {
public:
    int search(vector<int>& nums, int target) {
        int size = nums.size();
        int low = 0, high = size - 1, mid = (low + high) / 2;
        int ret = -1;
        while (low <= high) {
            // cout << nums[low] << ' ' << nums[mid] << ' ' << nums[high] << endl;
            if (nums[low] == target) return low;
            if (nums[high] == target) return high;
            if (nums[mid] == target) return mid;
            if (nums[mid] < target) {
                if (nums[mid] < nums[high] && nums[high] < target) {
                    high = mid - 1;
                } else {
                    low = mid + 1;
                }
            } else if (nums[mid] > target) {
                if (nums[mid] > nums[low] && nums[low] > target) {
                    low = mid + 1;
                } else {
                    high = mid - 1;
                }
            }
            mid = (low + high) / 2;
        }
        return ret;
    }
};

File: leetcode/python/0033-search-in-rotated-sorted-array.py
# https://leetcode.com/problems/search-in-rotated-sorted-array/
# Medium

# Check if nums[lo] <= nums[mid]
#   Then check if target is between nums[lo] and nums[mid]
#   And reduce hi accordingly
# Else
#   Check if target is between nums[mid] and nums[hi]
#   If it is then increase lo accordingly
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1
        
        while lo <= hi:
            mid = (lo + hi) // 2
            if target == nums[mid]:
                return mid
            
            if nums[lo] <= nums[mid]:
                if nums[lo] <= target <= nums[mid]:
                    hi = mid - 1
                else:
                    lo = mid + 1
            else:
                if nums[mid] <= target <= nums[hi]:
                    lo = mid + 1
                else:
                    hi = mid - 1
        return -1
                
        